ShvetsBy/rss-aws-be
# 10. Backend For Frontend

Task: https://github.com/rolling-scopes-school/cloud-development-course-initial/blob/main/10_backend_for_frontend/task.md
implementation: 
Get products http://shvetsby-bff-api-develop.eu-west-1.elasticbeanstalk.com/products
Get product http://shvetsby-bff-api-develop.eu-west-1.elasticbeanstalk.com/products/05e96ded-f2df-4743-9097-0539b6d65387
Post product http://shvetsby-bff-api-develop.eu-west-1.elasticbeanstalk.com/products

body for postman:
_{ 
"title": { "S": "Keychron beanslt Testkeeb"},
"description": { "S":"Wireless Mechanical Keyboard"},
"price": { "N": "59"},
"img": { "S": "KeychronK86"},
"count": { "N": "103" 
}_

Get cart http://shvetsby-bff-api-develop.eu-west-1.elasticbeanstalk.com/cart
deadline: 13.11
Score: 7/7

### Scope

- [x] 5 - The BFF Service should be deployed with Elastic Beanstalk. The BFF Service call should be redirected to the appropriate service : Product Service or Cart Service. The response from the BFF Service should be the same as if Product Service or Cart Service services were called directly.

1. You can check get and post methods for Products API https://h4uuydbl93.execute-api.eu-west-1.amazonaws.com/products
Use this object to check how method post works:
_{ 
"title": { "S": "Keychron Testkeeb"},
"description": { "S":"Wireless Mechanical Keyboard"},
"price": { "N": "59"},
"img": { "S": "KeychronK86"},
"count": { "N": "104" 
}_

single product https://h4uuydbl93.execute-api.eu-west-1.amazonaws.com/products/05e96ded-f2df-4743-9097-0539b6d65387

Also you can run get for Cart API http://shvetsby-cart-api-dev.eu-west-1.elasticbeanstalk.com/api/profile/cart

2. The same result you can get by running locally branch task-10 from this repo. 

- clone repo
- go to bff-service
- add .env file with products="https://h4uuydbl93.execute-api.eu-west-1.amazonaws.com"
cart="http://shvetsby-cart-api-dev.eu-west-1.elasticbeanstalk.com/api/profile"
- install dependencies
- run ./bin/www
- open postman and send:
1. get and post requests for http://localhost:3000/products 
2. get for single products like http://localhost:3000/products/ac9523ea-b7b7-4b43-8f03-80618dc7324b
3. get for http://localhost:3000/cart

Result will be the same like on this screens.
<img width="1282" alt="Снимок экрана 2022-11-09 в 18 45 17" src="https://user-images.githubusercontent.com/24583617/200875624-0e9df8f3-d25d-4f68-8223-d022d11a79bd.png">
<img width="1282" alt="Снимок экрана 2022-11-09 в 18 44 42" src="https://user-images.githubusercontent.com/24583617/200875638-ced09b8e-0a9c-4998-8415-6f6ed448053a.png">
<img width="1282" alt="Снимок экрана 2022-11-09 в 18 44 05" src="https://user-images.githubusercontent.com/24583617/200875658-9f3480e4-757e-49cf-9a06-e8a3a4802a40.png">
<img width="1282" alt="Снимок экрана 2022-11-10 в 12 04 32" src="https://user-images.githubusercontent.com/24583617/201047686-a8bd5231-3a83-48a9-9152-36c65c974940.png">


### Advanced

- [x] +1 - Add a cache at the BFF Service level for a request to the getProductsList lambda function of the Product Service. The cache should expire in 2 minutes.

How to test locally:
- Get products list
- Create new product
- Get products list - result shouldn’t have new product
- Wait more than 2 minutes
- Get products list - result should have new product

Task branch https://github.com/ShvetsBy/rss-aws-be/tree/task-10-nestjs
endpoint to check:
http://shvetsby-bff-api-dev.eu-west-1.elasticbeanstalk.com/products
http://shvetsby-bff-api-dev.eu-west-1.elasticbeanstalk.com/products/33fa8ef2-aa5d-45fb-8294-17ffd1f0e46e
https://shvetsby-bff-api-dev.eu-west-1.elasticbeanstalk.com/cart
